src/util/algorithms.cpp

```diff
@@ -0,0 +1,42 @@
+#include <util/algorithms.h>
+#include <goto-programs/goto_loops.h>
+#include <goto-programs/remove_skip.h>
+bool goto_functions_algorithm::run()
+{
+  Forall_goto_functions(it, goto_functions)
+  {
+    number_of_functions++;
+    runOnFunction(*it);
+    if(it->second.body_available)
+    {
+      goto_loopst goto_loops(it->first, goto_functions, it->second, this->msg);
+      auto function_loops = goto_loops.get_loops();
+      number_of_loops += function_loops.size();
+      if(function_loops.size())
+      {
+        goto_functiont &goto_function = it->second;
+        goto_programt &goto_program = goto_function.body;
+
+        // Foreach loop in the function
+        for(auto itt = function_loops.rbegin(); itt != function_loops.rend();
+            ++itt)
+          runOnLoop(*itt, goto_program);
+      }
+    }
+  }
+  goto_functions.update();
+  if(has_sideeffect())
+    remove_skip(goto_functions);
```

Comment:
I don't think you need to check for side-effects or to run remove_skip here. Since I couldn't find any place you are introducing skips.

Did you try just running update here?

Reply:
I think I've misunderstood what the `remove_skip` was intended for. I will change this.